### PR TITLE
chore(a3p): replace waitForBlock with retryUntilCondition on acceptance tests

### PR DESCRIPTION
**Pull request by Jorge-Lopes**
closes: https://github.com/Agoric/BytePitchPartnerEng/issues/10

## Description

This PR intends to replace the use of the `waitForBlock` method with the available `sync tools`, more specifically, the `retryUntilCondition` method.

The tests that were updated are:

- wallet.test.js
- core-eval.test.js
- value.Vow.test.js
- localchain.test.js

Note:

At wallet.test.js, the script `exitOffer.js` was replaced with the `broadcastBridgeAction` method of `walletUtils` to execute a `tryExitOffer` call.

### Security Considerations

No considerations.

### Scaling Considerations

No considerations.

### Documentation Considerations

No considerations.

### Testing Considerations

To build a condition that verifies the success of a `tryExitOffer` transaction for a failed or "bad invitation" offer, it was not possible to use the users wallet `liveOffers` due to the fact that when an invalid offer is created, the payment makes it to `liveOfferPayments` via `withdrawGive` but the offer fails to enter `liveOffers` or `liveOfferSeats`. 

Checking the Vstorage node `published.wallet.{addr}` for the status of `tryExitOffer` call also does not provide the needed information due to the error thrown by `tryExitOffer`, as the `offerId` isn’t present at `liveOfferSeats`.  

The alternative approach used was to monitor wallet balance, since the main goal is to ensure that the user’s payment is reclaimed,

Suggestion:  
If feasible, consider updating `Vstorage` to log a `tryExitOffer` attempt, even if it throws an error due to a missing `offerId`. This could include additional information about the transaction's outcome, such as “reclaim successful”.
This may require changes to `Vstorage` but could provide more transparent access to the outcome of similar scenarios.

This topic was addressed in the following discussion https://github.com/Agoric/agoric-sdk/discussions/10374

### Upgrade Considerations

No considerations.

### CODE REVIEW

File: a3p-integration/proposals/z:acceptance/localchain.test.js [hunk 1]
```diff
@@ -27,7 +29,12 @@ test(`localchain passes tests`, async t => {
   const nodePath = 'test.localchain';
   const nodeValue = JSON.stringify({ success: true });
 
-  await waitForBlock(2); // enough time for core eval to execute ?
+  await retryUntilCondition(
+    async () => readPublished(nodePath),
+    value => value === nodeValue,
+    'core eval not enforced yet',
```
Comment (Chris-Hibbert) [inline a3p-integration/proposals/z:acceptance/localchain.test.js]:
```suggestion
    'core eval not processed yet',
```

File: a3p-integration/proposals/z:acceptance/localchain.test.js [hunk 2]
```diff
@@ -27,7 +29,12 @@ test(`localchain passes tests`, async t => {
   const nodePath = 'test.localchain';
   const nodeValue = JSON.stringify({ success: true });
 
-  await waitForBlock(2); // enough time for core eval to execute ?
+  await retryUntilCondition(
```
Comment (Chris-Hibbert) [inline a3p-integration/proposals/z:acceptance/localchain.test.js]: `retryUntilCondition` returns the produced value.
```suggestion
  const value = await retryUntilCondition(
```

Line 39 doesn't have to `await` it again.

File: a3p-integration/proposals/z:acceptance/valueVow.test.js [hunk 3]
```diff
@@ -44,7 +44,14 @@ test('vow survives restart', async t => {
   });
 
   t.log('confirm the value is not in offer results');
-  await waitForBlock(2);
+  await retryUntilCondition(
```
Comment (Chris-Hibbert) [inline a3p-integration/proposals/z:acceptance/valueVow.test.js]:
```suggestion
  const getterStatus = await retryUntilCondition(
```

File: a3p-integration/proposals/z:acceptance/wallet.test.js [hunk 4]
```diff
@@ -43,20 +52,25 @@ test.serial(`send invitation via namesByAddress`, async t => {
...
+  });
 
-  await $`node ./exitOffer.js --id ${offerId} --from ${from} `;
-  await waitForBlock(2);
+  await retryUntilCondition(
+    async () => getBalances([GOV1ADDR], 'uist'),
+    istBalanceAfter => istBalanceAfter > istBalanceBefore,
+    'tryExitOffer failed to reclaim stuck payment ',
+    { setTimeout, retryIntervalMs: 5000, maxRetries: 15 },
+  );
```
Comment (Chris-Hibbert) [inline a3p-integration/proposals/z:acceptance/wallet.test.js]:
```suggestion
    const istBalanceAfter = await retryUntilCondition(
    async () => getBalances([GOV1ADDR], 'uist'),
    istBalanceAfter => istBalanceAfter > istBalanceBefore,
    'tryExitOffer failed to reclaim stuck payment ',
    { setTimeout, retryIntervalMs: 5000, maxRetries: 15 },
  );
 ```

**Review by Chris-Hibbert — APPROVED** [review]
Looks fine. My only suggestion (repeatedly) was to use the return value from `retryUntilCondition()`.

File: a3p-integration/proposals/z:acceptance/localchain.test.js (hunk 1 above)
Comment (Jorge-Lopes) [reply a3p-integration/proposals/z:acceptance/localchain.test.js]: Requested change addressed

File: a3p-integration/proposals/z:acceptance/localchain.test.js (hunk 2 above)
Comment (Jorge-Lopes) [reply a3p-integration/proposals/z:acceptance/localchain.test.js]: Requested change addressed

File: a3p-integration/proposals/z:acceptance/wallet.test.js (hunk 4 above)
Comment (Jorge-Lopes) [reply a3p-integration/proposals/z:acceptance/wallet.test.js]: Requested change addressed

File: a3p-integration/proposals/z:acceptance/valueVow.test.js (hunk 3 above)
Comment (Jorge-Lopes) [reply a3p-integration/proposals/z:acceptance/valueVow.test.js]: Requested change addressed

**Comment by Jorge-Lopes** [comment]
Thank you for the review @Chris-Hibbert .
Regarding the testing considerations, do you have feedback on a possible approach for handling tryExitOffer transactions visibility?

**Comment by Chris-Hibbert** [comment]
> Regarding the testing considerations, do you have feedback on a possible approach for handling tryExitOffer transactions visibility?

I think monitoring the wallet balance is the best we can do. This is an exception condition, which makes it harder to log status. I think if Zoe blocks the offer, the wallet never hears about it, so there's no place for it to be aware of the attempt. It's acceptable that exception cases sometimes take more work to monitor. The normal cases now have good visibility.

**Pull request merged**